tests/ui/mem_replace.rs

```diff
@@ -9,13 +9,30 @@
 
 // run-rustfix
 #![allow(unused_imports)]
-#![warn(clippy::all, clippy::style, clippy::mem_replace_option_with_none)]
+#![warn(
+    clippy::all,
+    clippy::style,
+    clippy::mem_replace_option_with_none,
+    clippy::mem_replace_with_default
+)]
 
 use std::mem;
 
-fn main() {
```

Comment:
Could you add macro test cases?

Reply:
Hey could you give me an example of what a good macro test case would look like for this? Are we checking if the lint is enforced or not?

Reply:
I am not sure what is the best test case. But here is one I can think of:
```rust
macro_rules! foo {
    ($s:expr) => {
        std::mem::replace(&mut $s, Default::default())
    };
}

pub fn bar() {
    let mut s = String::from("bar");
    foo!(s);
}
```

Reply:
Hey do you think it's better to not lint within macros?

The lint works in a macro and gives the error:
```
error: replacing a value of type `T` with `T::default()` is better expressed using `std::mem::take`
  --> $DIR/mem_replace.rs:23:9
   |
LL |         std::mem::replace($s, Default::default())
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using: `std::mem::take(s)`
...
LL |     rwd!(s);
   |     -------- in this macro invocation
```

But the suggestion is incorrect(needs to be `std::mem::take($s)` instead of the variable name `s`) and I am not sure how to get the macro argument from the span. Also, looks like linting within macros can result in a lot of false positives so I am a little hesistant. Would love to know your thoughts.

Reply:
I would skip the macro cases too. 
Feel free to hear from other clippy people @phansch and @flip1995

Reply:
I think linting local macros is ok, external macros is problematic. So we should exclude external macros. Since you correctly use `snippet_with_applicability` this won't be autofixed by rustfix, so we're good at that front.

Reply:
@flip1995 Hey so I allowed the lint for internal macros but even with `snippet_with_applicability` the fixed file is updated inside a macro so I explicitly set the applicability to `MaybeIncorrect` when inside a macro as shown below but that didn't work either:

```rust
let applicability = if in_macro(expr.span) {
                Applicability::MaybeIncorrect
} else {
                Applicability::MachineApplicable
};

span_lint_and_sugg(
                cx,
                MEM_REPLACE_WITH_DEFAULT,
                expr.span,
                "replacing a value of type `T` with `T::default()` is better expressed using `std::mem::take`",
                "consider using",
                format!("std::mem::take({})", snippet(cx, args[0].span, "")),
                applicability,
);
```

rustfix still generates:
```
error: replacing a value of type `T` with `T::default()` is better expressed using `std::mem::take`
  --> $DIR/mem_replace.rs:27:9
   |
LL |         std::mem::replace($s, Default::default())
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using: `std::mem::take(s)`
...
LL |     take!(s);
   |     --------- in this macro invocation
```
and the fixed code is incorrect because the variable name is wrong. Any thoughts?

Reply:
When writing `// run-rustfix` on top of the file, rustfix will exchange **every** suggestion, no matter the applicability. There is the additional comment `// rustfix-only-machine-applicable` which only applies machine applicable suggestions.